### PR TITLE
[settings-view] Update package snippets view to reflect new features

### PR DESCRIPTION
**Pull request by savetheclocktower**
### tl;dr

The snippets listing table will now show command names (if any are present), and copying the snippet will preserve any `prefix` or `command` property (or both).

Snippet scope names are now represented exactly as they were described in the original snippet file.

### Identify the Bug

Here’s an existing bug with this view: the scope name shown is a normalized version that’s a bit… weird. For example, even if you specify `.source.js` in your snippets file, it’ll alphabetize the selector to `.js.source` — technically equivalent, but off-putting.

But that’s not the bug: the bug is that you’re allowed to specify a comma-separated scope, but the snippets view doesn’t represent them properly in the table. For instance, the `language-javascript` package’s snippets are all defined under `.source.js, .source.flow`, but the package settings screen lists them all under the scope `.flow.source`. If someone wants to tweak an existing snippet, we tell them to hit the <kbd>Copy</kbd> button and paste it into their snippets file, but that won’t work unless it captures the full intended scope.

The other “bugs” are things that need fixing now that snippets can do more things. For instance, that same <kbd>Copy</kbd> button should preserve the `command` property, if it exists, and should only add a `prefix` property if the snippet has one. And now that a snippet can have a `command` property, it should show up in the table for discoverability.

### Description of the Change

As part of [snippets#10](https://github.com/pulsar-edit/snippets/pull/10), I made it so that a snippet’s metadata object includes a `selector` property that preserves the original scope name under which the snippet was defined. I didn’t even realize that that would come in handy on the settings-view side, and honestly it’s how I should’ve implemented this [way back when](https://github.com/atom/settings-view/pull/1084).

I added a “Command” column in the table, but I’m quite aware of how few packages will provide snippets that are bound to commands (at least at first); hence I made it so that the `th` and all its corresponding `td`s hide themselves if _no_ snippets have `command` defined. That way they won’t just hog a bunch of empty table space.

### Alternate Designs

There’s something I went back and forth on: at first, I thought that it made more sense for the “Enable” checkbox on the snippets view to control only the tab triggers. After all, there’s not so much of a need to disable a snippet when it’s only bound to a command and not a prefix — you won’t really trigger it by accident, and it won’t get suggested to you by `autocomplete-snippets`. And if the package maps it to a key and you don’t like that, you can disable or unset the keymap itself.

But I realized it would be a pretty big architectural disruption for snippet disabling to behave that way, on top of which it would probably confuse users. If a user wants to disable snippets but re-enable the ones that define commands, they can do so through the <kbd>Copy</kbd> button and an override in their `snippets.cson`.

### Possible Drawbacks

Other than what I just described, I can’t think of any.

### Verification Process

I wrote new specs, but they’re mostly behind a version guard, because the associated behavior change in `snippets` won’t land until 1.103 is released. (I probably should be using the rolling version locally, but I haven’t gotten around to it.) I’m submitting this PR because I was compelled to work on it while I had the thought, but we should wait until the new stuff goes out in a standard release.

### Release Notes

- The settings-view package now lists a package’s snippets more accurately.

### CODE REVIEW

**Comment by savetheclocktower** [comment]
Also, I just realized that the rolling release won't have the new snippets stuff from [snippets#10](https://github.com/pulsar-edit/snippets/pull/10) because the version hasn't been bumped in this repo yet. That's a blocker for the new specs passing, even if I remove the version guard.

**Comment by savetheclocktower** [comment]
A taste of what this view now looks like; forgive my weird UI theme.

Screen Shot 2023-02-26 at 11 12 31 PM

**Comment by confused-Techie** [comment]
> Also, I just realized that the rolling release won't have the new snippets stuff from [snippets#10](https://github.com/pulsar-edit/snippets/pull/10) because the version of `snippets` hasn't been bumped in this repo yet. That's a blocker for the new specs passing, even if I remove the version guard.

Might not be a bad idea to get a PR in to bump the version of snippets here, and add that as a blocker to your initial PR description.

If you do go ahead and do this, I'd like to please request to include the PR's that'll be included with this bump.

Otherwise I'll try to get to it when possible this week

**Comment by savetheclocktower** [comment]
Oh, also: I decaffeinated the spec that I touched. The tests I changed were under `"displays the snippets registered by the package"`; look for the version guard against `1.103`.

**Comment by savetheclocktower** [comment]
> If you do go ahead and do this, I'd like to please request to include the PR's that'll be included with this bump.

Do you mean that I should specify in the bump-PR description a list of PRs that have landed since the last version bump of `snippets`?

**Comment by confused-Techie** [comment]
> > If you do go ahead and do this, I'd like to please request to include the PR's that'll be included with this bump.
> 
> Do you mean that I should specify in the bump-PR description a list of PRs that have landed since the last version bump of `snippets`?

Yeah exactly, a good example would be how we handled it in #383 

It just makes writing the ChangeLogs and reviewing it super easy

**Review by DeeDeeG — COMMENTED** [review]
This should be fully un-blocked now that https://github.com/pulsar-edit/pulsar/pull/408 has been merged.

I read the description, this sounds really neat and helpful. Nothing too crazy, per the description at least.

I like that the diff is pretty small for the number of files it touches, that's great to see!

... Other than the spec decaf, in theory it'd be great to do that separately, maybe in its own PR, but at least in its own commit. However, we may be able to take this as-is if someone goes in and verifies that the diff (compared to the plain "just decaffed" version of the file) looks good.

Again, I'm very unfamiliar with snippets, so I don't think I can confidently review the technical changes, but I do appreciate the intent and I don't outright see anything to object to, just that I don't totally know what I'm looking at here, personally speaking.

File: packages/settings-view/spec/installed-package-view-spec.js [hunk 1]
```diff
@@ -0,0 +1,517 @@
...
+          view.element.ownerDocument.querySelector('.snippet-body-tooltip')
+        ).toExist();
+      });
+    })
+  );
+
+
+  // Relies on behavior not present in the snippets package before 1.33.
+  // TODO: These tests should always run once 1.33 is released.
+  if (parseFloat(atom.getVersion()) >= 1.33) {
```
Comment (confused-Techie) [inline packages/settings-view/spec/installed-package-view-spec.js:203]: We may want to remove this check as a whole. Since I remember we had [issues when Hydrogen](https://github.com/nteract/hydrogen/pull/2162) was using `parseFloat` to check versions once we starting changing the version of Pulsar. So we might want to remove this if statement as a whole. This isn't a critique of your changes, just if we are already here might be a good idea, especially since we no longer need to even consider these old versions.

**Review by confused-Techie — COMMENTED** [review]
Overall like @DeeDeeG mentioned it was a good idea to check the decafe work.

The decafe work here looks great! I went ahead and grabbed the CoffeeScript source, and got the machine decaf changes, then compared the diff between the machine decaf and your changes. Overall everything looks great, with the actual decaf work being very subtle and focused mainly on using arrow functions where appropriate and using proper returns (ie adding returns or excluded as needed).

I will admit it is a little dicey comparing the diff when there is decafe work and new changes to the tests. But to the best of my ability these look good, and I'd be happy merging the changes in `installed-package-view.spec.js` as is without much change (although did make a quick comment on one part).

But otherwise hope to getting to review the beef of content soon, this looks great, thanks!

**Comment by savetheclocktower** [comment]
My instinct in decaffeinating `installed-package-view-spec.js` was that it would be weird to touch a CoffeeScript file and not decaffeinate it, but I do understand how it made the diff harder to understand. I'll keep that in mind for next time.

**Comment by mauricioszabo** [comment]
@savetheclocktower can you add an entry on CHANGELOG.md, just below the `## [Unrelease]` line, explaining the changes on snippets?

**Review by confused-Techie — APPROVED** [review]
Lets get this one merged!
Thanks a ton as always @savetheclocktower

**Comment by confused-Techie** [comment]
Alright, this has been up some time with no further development, I'll go ahead and merge this one.

Thanks a ton @savetheclocktower for everything here
